File: backend/app/schemas/daily_metrics.py
from __future__ import annotations

from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import date, datetime


class DailyMetricsBase(BaseModel):
    date: date
    weight_kg: Optional[float] = None
    body_fat_percentage: Optional[float] = None
    muscle_mass_kg: Optional[float] = None
    sleep_duration_hours: Optional[float] = None
    sleep_quality_score: Optional[int] = Field(None, ge=1, le=10)
    bedtime: Optional[str] = None
    wake_time: Optional[str] = None
    fatigue_level: Optional[int] = Field(None, ge=1, le=10)
    motivation_level: Optional[int] = Field(None, ge=1, le=10)
    stress_level: Optional[int] = Field(None, ge=1, le=10)
    energy_level: Optional[int] = Field(None, ge=1, le=10)
    training_readiness: Optional[int] = Field(None, ge=1, le=10)
    recovery_status: Optional[str] = None
    resting_heart_rate: Optional[int] = None
    blood_pressure_systolic: Optional[int] = None
    blood_pressure_diastolic: Optional[int] = None
    notes: Optional[str] = None
    mood_tags: Optional[List[str]] = None
    is_estimated: Optional[bool] = False
    data_source: Optional[str] = None

    class Config:
        from_attributes = True
        arbitrary_types_allowed = True
        validate_assignment = False


class DailyMetricsCreate(DailyMetricsBase):
    pass


class DailyMetricsUpdate(BaseModel):
    weight_kg: Optional[float] = None
    body_fat_percentage: Optional[float] = None
    muscle_mass_kg: Optional[float] = None
    sleep_duration_hours: Optional[float] = None
    sleep_quality_score: Optional[int] = Field(None, ge=1, le=10)
    bedtime: Optional[str] = None
    wake_time: Optional[str] = None
    fatigue_level: Optional[int] = Field(None, ge=1, le=10)
    motivation_level: Optional[int] = Field(None, ge=1, le=10)
    stress_level: Optional[int] = Field(None, ge=1, le=10)
    energy_level: Optional[int] = Field(None, ge=1, le=10)
    training_readiness: Optional[int] = Field(None, ge=1, le=10)
    recovery_status: Optional[str] = None
    resting_heart_rate: Optional[int] = None
    blood_pressure_systolic: Optional[int] = None
    blood_pressure_diastolic: Optional[int] = None
    notes: Optional[str] = None
    mood_tags: Optional[List[str]] = None
    is_estimated: Optional[bool] = None
    data_source: Optional[str] = None


class DailyMetricsResponse(DailyMetricsBase):
    id: str
    user_id: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True
        arbitrary_types_allowed = True
        validate_assignment = False


class WeeklyMetricsSummaryBase(BaseModel):
    week_start_date: date
    week_end_date: date
    avg_weight: Optional[float] = None
    avg_body_fat_percentage: Optional[float] = None
    avg_muscle_mass_kg: Optional[float] = None
    avg_sleep_hours: Optional[float] = None
    avg_sleep_quality: Optional[float] = None
    avg_fatigue_level: Optional[float] = None
    avg_motivation_level: Optional[float] = None
    avg_stress_level: Optional[float] = None
    avg_energy_level: Optional[float] = None
    avg_training_readiness: Optional[float] = None
    avg_resting_heart_rate: Optional[float] = None
    total_entries: int = 0

    class Config:
        from_attributes = True
        arbitrary_types_allowed = True
        validate_assignment = False


class WeeklyMetricsSummaryResponse(WeeklyMetricsSummaryBase):
    id: str
    user_id: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True
        arbitrary_types_allowed = True
        validate_assignment = False


class MonthlyMetricsSummaryBase(BaseModel):
    year: int
    month: int
    avg_weight: Optional[float] = None
    avg_body_fat_percentage: Optional[float] = None
    avg_muscle_mass_kg: Optional[float] = None
    avg_sleep_hours: Optional[float] = None
    avg_sleep_quality: Optional[float] = None
    avg_fatigue_level: Optional[float] = None
    avg_motivation_level: Optional[float] = None
    avg_stress_level: Optional[float] = None
    avg_energy_level: Optional[float] = None
    avg_training_readiness: Optional[float] = None
    avg_resting_heart_rate: Optional[float] = None
    total_entries: int = 0

    class Config:
        from_attributes = True
        arbitrary_types_allowed = True
        validate_assignment = False


class MonthlyMetricsSummaryResponse(MonthlyMetricsSummaryBase):
    id: str
    user_id: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True
        arbitrary_types_allowed = True
        validate_assignment = False


class DailyMetricsListResponse(BaseModel):
    items: List[DailyMetricsResponse]
    total: int
    page: int
    limit: int
    total_pages: int

    class Config:
        from_attributes = True
        arbitrary_types_allowed = True
        validate_assignment = False


class MetricsTrendResponse(BaseModel):
    metric_name: str
    trend_data: List[Dict[str, Any]]
    trend_direction: str
    change_percentage: float
    period: str
    confidence: Optional[float] = None

    class Config:
        from_attributes = True
        arbitrary_types_allowed = True
        validate_assignment = False


class HealthInsightsResponse(BaseModel):
    insight_type: str
    message: str
    severity: str
    recommendations: List[str]
    confidence: Optional[float] = None
    related_metrics: Optional[List[str]] = None

    class Config:
        from_attributes = True
        arbitrary_types_allowed = True
        validate_assignment = False


class MetricsComparisonResponse(BaseModel):
    current_period: Dict[str, Any]
    previous_period: Dict[str, Any]
    changes: Dict[str, float]
    period_type: str

    class Config:
        from_attributes = True
        arbitrary_types_allowed = True
        validate_assignment = False

File: frontend-react/src/types/dailyMetrics.ts
export interface DailyMetricsBase {
  date: string
  
  // 身体データ
  weight_kg?: number
  body_fat_percentage?: number
  muscle_mass_kg?: number
  
  // 睡眠データ
  sleep_duration_hours?: number
  sleep_quality_score?: number
  bedtime?: string
  wake_time?: string
  
  // コンディション評価
  fatigue_level?: number
  motivation_level?: number
  stress_level?: number
  energy_level?: number
  
  // 運動関連
  training_readiness?: number
  recovery_status?: 'excellent' | 'good' | 'fair' | 'poor'
  
  // 健康指標
  resting_heart_rate?: number
  blood_pressure_systolic?: number
  blood_pressure_diastolic?: number
  
  // メモ・その他
  notes?: string
  mood_tags?: string[]
}

export interface DailyMetricsCreate extends DailyMetricsBase {}

export interface DailyMetricsUpdate {
  weight_kg?: number
  body_fat_percentage?: number
  muscle_mass_kg?: number
  sleep_duration_hours?: number
  sleep_quality_score?: number
  bedtime?: string
  wake_time?: string
  fatigue_level?: number
  motivation_level?: number
  stress_level?: number
  energy_level?: number
  training_readiness?: number
  recovery_status?: 'excellent' | 'good' | 'fair' | 'poor'
  resting_heart_rate?: number
  blood_pressure_systolic?: number
  blood_pressure_diastolic?: number
  notes?: string
  mood_tags?: string[]
}

export interface DailyMetricsResponse extends DailyMetricsBase {
  id: string
  user_id: string
  is_estimated: boolean
  data_source?: string
  created_at: string
  updated_at: string
}

export interface WeeklyMetricsSummaryResponse {
  id: string
  user_id: string
  week_start_date: string
  week_end_date: string
  
  // 週間平均値
  avg_weight_kg?: number
  avg_sleep_duration_hours?: number
  avg_fatigue_level?: number
  avg_motivation_level?: number
  avg_stress_level?: number
  avg_energy_level?: number
  avg_training_readiness?: number
  avg_resting_heart_rate?: number
  
  // 週間トレンド
  weight_trend?: string
  sleep_trend?: string
  fatigue_trend?: string
  motivation_trend?: string
  
  // データ品質
  data_completeness?: number
  days_recorded?: number
  
  created_at: string
  updated_at: string
}

export interface MonthlyMetricsSummaryResponse {
  id: string
  user_id: string
  year: number
  month: number
  
  // 月間平均値
  avg_weight_kg?: number
  avg_sleep_duration_hours?: number
  avg_fatigue_level?: number
  avg_motivation_level?: number
  avg_stress_level?: number
  avg_energy_level?: number
  avg_training_readiness?: number
  avg_resting_heart_rate?: number
  
  // 月間統計
  weight_change_kg?: number
  sleep_consistency_score?: number
  stress_peak_days?: number
  low_energy_days?: number
  
  // データ品質
  data_completeness?: number
  days_recorded?: number
  
  created_at: string
  updated_at: string
}

export interface DailyMetricsListResponse {
  items: DailyMetricsResponse[]
  total: number
  page: number
  limit: number
  total_pages: number
}

export interface MetricsTrendResponse {
  dates: string[]
  weight_kg: (number | null)[]
  sleep_duration_hours: (number | null)[]
  fatigue_level: (number | null)[]
  motivation_level: (number | null)[]
  stress_level: (number | null)[]
  energy_level: (number | null)[]
  training_readiness: (number | null)[]
  resting_heart_rate: (number | null)[]
}

export interface HealthInsightsResponse {
  overall_health_score: number
  sleep_quality_score: number
  stress_management_score: number
  recovery_score: number
  recommendations: string[]
  trends: Record<string, any>
  alerts: string[]
}
